### PR TITLE
ipatests: Test for OTP when the LDAP connection timed out

### PR DESCRIPTION
**Pull request by amore17**
Test to verify that when the nsslapd-idletimeout is exceeded (50s idle, 60s sleep) 
then the ipa-otpd process should exit without error.

Test for : https://pagure.io/freeipa/issue/6587

### CODE REVIEW

File: ipatests/test_integration/test_otp.py [hunk 1]
```diff
@@ -309,3 +311,51 @@ def test_2fa_disable_single_prompt(self):
...
+        """Test for ipa-otpd when dirsrv is restart with OTP auth.
+
+        Test for : https://pagure.io/freeipa/issue/6587
+
+        When dirsrv is restarted and there is OTP authentication, then after
+        dirsrv restart there should be no failure in service ipa-otpd.
+        """
+        since = time.strftime('%H:%M:%S')
+        master = self.master
+        tasks.kinit_admin(self.master)
```
Comment (mrizwan93) [inline ipatests/test_integration/test_otp.py:325]: replace `self.master` with master or use `self.master` at all the places.

File: ipatests/test_integration/test_otp.py [hunk 2]
```diff
@@ -309,3 +311,51 @@ def test_2fa_disable_single_prompt(self):
...
             self.master.run_command(['semanage', 'login', '-D'])
             sssd_conf_backup.restore()
+
+    def test_check_otpd_after_dirsrv_restart(self):
+        """Test for ipa-otpd when dirsrv is restart with OTP auth.
+
+        Test for : https://pagure.io/freeipa/issue/6587
+
+        When dirsrv is restarted and there is OTP authentication, then after
+        dirsrv restart there should be no failure in service ipa-otpd.
```
Comment (mrizwan93) [inline ipatests/test_integration/test_otp.py:321]: Description can be simplified like:
When OTP authentication is set and dirsrv restrted, ipa-otpd service should not fail.

File: ipatests/test_integration/test_otp.py [hunk 3]
```diff
@@ -309,3 +311,51 @@ def test_2fa_disable_single_prompt(self):
...
+            result1 = master.run_command(
+                ['systemctl', 'list-units', '--state=failed'])
+            assert "ipa-otpd" not in result1.stdout_text
+            result2 = master.run_command(
+                ['journalctl', '--since={}'.format(since)])
+            regex = r".*ipa-otpd@.*\sSucceeded"
+            assert re.search(regex, result2.stdout_text)
+        finally:
+            del_otptoken(master, otpuid)
+            tasks.ldapmodify_dm(master, ldap_query.format(limit=orig_limit))
```
Comment (mrizwan93) [inline ipatests/test_integration/test_otp.py:361]: I would move setup and removal part of otp token in fixture.

**Review by mrizwan93 — COMMENTED** [review]
@amore17 Thanks for the patch. Please find inline comments

**Comment by amore17** [comment]
[logs](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/b52899a8-ee09-11eb-be63-525400604882)

**Review by mrizwan93 — APPROVED** [review]
LGTM.

File: ipatests/test_integration/test_otp.py [hunk 4]
```diff
@@ -309,3 +311,58 @@ def test_2fa_disable_single_prompt(self):
...
+        ipa-otpd service should not fail.
+        """
+        since = time.strftime('%H:%M:%S')
+        tasks.kinit_admin(self.master)
+        otpuid, totp = add_otptoken(self.master, USER, otptype="totp")
+        try:
+            # restart dirsrv.target
+            serverid = realm_to_serverid(self.master.domain.realm)
+            self.master.run_command(
+                ['systemctl', 'restart', 'dirsrv@%s.service' % serverid]
```
Comment (rcritten) [inline ipatests/test_integration/test_otp.py:351]: I'm not sure that a restart is required.

File: ipatests/test_integration/test_otp.py [hunk 5]
```diff
@@ -309,3 +311,58 @@ def test_2fa_disable_single_prompt(self):
...
+            )
+
+            # kinit with OTP auth
+            otpvalue = totp.generate(int(time.time())).decode("ascii")
+            kinit_otp(self.master, USER, password=PASSWORD, otp=otpvalue)
+            time.sleep(60)
+            failed_services = self.master.run_command(
+                ['systemctl', 'list-units', '--state=failed']
+            )
```
Comment (rcritten) [inline ipatests/test_integration/test_otp.py:360]: I wonder if you should run this first to be sure there are no failed units when the test starts.

File: ipatests/test_integration/test_otp.py [hunk 6]
```diff
@@ -309,3 +311,58 @@ def test_2fa_disable_single_prompt(self):
...
+        # cleanup
+        tasks.ldapmodify_dm(self.master, ldap_query.format(limit=orig_limit))
+
+    def test_check_otpd_after_dirsrv_restart(self, setup_otp_nsslapd):
+        """Test for ipa-otpd when dirsrv is restart with OTP auth.
+
+        Test for : https://pagure.io/freeipa/issue/6587
+
+        When OTP authentication is set and dirsrv restarted,
+        ipa-otpd service should not fail.
```
Comment (rcritten) [inline ipatests/test_integration/test_otp.py:342]: The issue was that the daemon was exiting with a failure when the LDAP connection timed out. This should have been, and is, handled more gracefully now. The test is ensuring that when the idle timeout is exceeded (50s idle, 60s sleep) then the ipa-otpd process exits without error.

File: ipatests/test_integration/test_otp.py (hunk 4 above)
Comment (amore17) [reply ipatests/test_integration/test_otp.py:351]: Not required.
reproducer logs : https://privatebin-it-iso.int.open.paas.redhat.com/?111647f526f8067a#EUJAkNAQbWx3gXH3eTaPJiNoExsrQufVF2pV9HER1iAN

Comment (amore17) [reply ipatests/test_integration/test_otp.py:351]: Not required.
[reproducer logs](https://privatebin-it-iso.int.open.paas.redhat.com/?cd75f7ac455a2a18#FP6qPfjcRE5sJnj7LJ6Wjqd82T5wPyTjjfSZUsf9yddY).

**Comment by amore17** [comment]
[logs](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/9e575ad0-eebc-11eb-911a-525400b889ed)

File: ipatests/test_integration/test_otp.py (hunk 4 above)
Comment (amore17) [reply ipatests/test_integration/test_otp.py]: Not required.
[reproducer logs](https://privatebin-it-iso.int.open.paas.redhat.com/?42c9af3031b7b352#Gcu4Qfh47MmjjAZHRzSqMQjvhNSnn4c2ADVz2Pfnuqws)

File: ipatests/test_integration/test_otp.py [hunk 7]
```diff
@@ -309,3 +310,58 @@ def test_2fa_disable_single_prompt(self):
...
+        check_services = self.master.run_command(
+            ['systemctl', 'list-units', '--state=failed']
+        )
+        assert "0 loaded units listed" in check_services.stdout_text
+        assert "ipa-otpd" not in check_services.stdout_text
+        yield
+        # cleanup
+        tasks.ldapmodify_dm(self.master, ldap_query.format(limit=orig_limit))
+
+    def test_check_otpd_after_dirsrv_restart(self, setup_otp_nsslapd):
```
Comment (rcritten) [inline ipatests/test_integration/test_otp.py:340]: Can you change this from restart to something like test_check_otpd_after_idle_timeout ?

File: ipatests/test_integration/test_otp.py [hunk 8]
```diff
@@ -309,3 +310,58 @@ def test_2fa_disable_single_prompt(self):
...
+        """Test for OTP when the LDAP connection timed out.
+
+        Test for : https://pagure.io/freeipa/issue/6587
+
+        ipa-otpd was exiting with failure when LDAP connection timed out.
+        Test to verify that when the nsslapd-idletimeout is exceeded (50s idle,
+        60s sleep) then the ipa-otpd process should exit without error.
```
Comment (rcritten) [inline ipatests/test_integration/test_otp.py:347]: You increased the idle to 60s and the timeout to 90s so it doesn't match the comment.

I don't know how much you want to play with this but a 30/45 second idle/sleep would also probably work and would ever so slightly speed up the test.

**Comment by rcritten** [comment]
The changes LGTM If it passes CI with the reduced idle timeout please remove the temp commit **and** update the commit message with the new times.

File: ipatests/test_integration/test_otp.py (hunk 8 above)
Comment (amore17) [reply ipatests/test_integration/test_otp.py:347]: Changed to 30/60 second idle/sleep.
```
Jul 27 15:51:50 master.ipa.test ipa-otpd[23662]: opttestuser@IPA.TEST: response sent: Access-Accept
Jul 27 15:52:50 master.ipa.test systemd[1]: ipa-otpd@1-21071-0.service: Succeeded.
```

Comment (amore17) [reply ipatests/test_integration/test_otp.py:347]: [logs](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/e6d75b64-eef0-11eb-bee7-fa163e6f7f4d/)

**Comment by amore17** [comment]
logs with [temp_commit](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/e6d75b64-eef0-11eb-bee7-fa163e6f7f4d/)

**Comment by rcritten** [comment]
Thanks, the changes look good. The commit has two signatures though, as if two commits were combined. Can you remove the extra one and I'll ack this?

**Comment by amore17** [comment]
> Thanks, the changes look good. The commit has two signatures though, as if two commits were combined. Can you remove the extra one and I'll ack this?

Updated.

**Comment by flo-renaud** [comment]
master:

* 81b615787768ec3fc2b55f4b49bd9515dc24a79d ipatests: Test for OTP when the LDAP connection timed out.

**Comment by flo-renaud** [comment]
@amore17 a manual backport is required for ipa-4-6 branch, can you handle it? Thanks

**Pull request closed without merging**